### PR TITLE
[JUJU-2116]: Store default-base in model-config

### PR DESCRIPTION
**Pull request by SimonRichardson**
The basic premise for this change is to start pushing `default-base` through the stack so that we no longer have to depend on `series`. This is one step towards the end goal. This has a lot of changes because we're now stating that we prefer a base over a series.

`default-series` is removed from model-config completely. We swap out the `default-series` when doing a migration import for a `default-base`. This means we only ever store a `default-base`, conversion between a base to a series is done at the API facade layer, and that is to provide backward compatibility. Once series is removed, we can then remove all those base to series conversions.

There is no synchronization happening, we're just using default-base as the source of truth and mandating that you can always go to a series. This keeps the data model clean.

```sh
$ juju model-config -m controller default-base=ubuntu@22.04
$ juju model-config -m controller default-series
jammy
$ juju model-config -m controller default-series=focal
$ juju model-config -m controller default-base
ubuntu@20.04
$ juju model-config -m controller default-base=""
$ juju model-config -m controller default-series

$ juju model-config -m controller default-base

$ juju model-config -m controller default-base=blah@blah
ERROR os "blah" version "blah" not found (not found)
```

Additional work in bootstrap was also required, as the model config forced changes at the CLI. Internally the boostrap args will use bases before it hits the environ bootstrap params. We can push bases through that at another PR.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

### Series not found:

```sh
$ juju bootstrap lxd test --build-agent --model-default default-series=foo
WARNING the default-series configuration option is deprecated, use default-base instead
ERROR series "foo" not supported
```

### Invalid base:

```sh
$ juju bootstrap lxd test --build-agent --bootstrap-base=foo
ERROR base "foo" not valid
```

### Invalid series:

```sh
$ juju bootstrap lxd test --build-agent --bootstrap-series=foo
ERROR cannot determine base for series "foo"
```

### Can not supply both default-series or default-base

```sh
$ juju bootstrap lxd test --build-agent --bootstrap-series=foo --bootstrap-base=bar
ERROR cannot specify both --bootstrap-series and --bootstrap-base
```

### Using a base

```sh
$ juju bootstrap lxd test --build-agent --bootstrap-base=ubuntu@22.04
```

### Using a series

```sh
$ juju bootstrap lxd test --build-agent --bootstrap-series=jammy
```

### Model config whilst bootstrapping:

```sh
$ juju bootstrap lxd test --build-agent --config default-series=focal
```

### Model default whilst bootstrapping:

```sh
$ juju bootstrap lxd test --build-agent --model-default default-series=focal
```

### Model config changes

```sh
$ juju model-config -m controller default-base=ubuntu@22.04
$ juju model-config -m controller default-series
jammy
$ juju model-config -m controller default-series=focal
$ juju model-config -m controller default-base
ubuntu@20.04
$ juju model-config -m controller default-base=""
$ juju model-config -m controller default-series

$ juju model-config -m controller default-base

$ juju model-config -m controller default-base=blah@blah
ERROR os "blah" version "blah" not found (not found)
```

### Deploying with model config

```sh
$ juju add-model default
$ juju model-config default-base=ubuntu@22.04
$ juju deploy ubuntu
Located charm "ubuntu" in charm-hub, revision 21
Deploying "ubuntu" from charm-hub charm "ubuntu", revision 21 in channel stable on ubuntu@22.04/stable
$ juju model-config default-series=focal
$ juju model-config default-base
ubuntu@20.04/stable
$ juju deploy ubuntu ubuntux
Located charm "ubuntu" in charm-hub, revision 21
Deploying "ubuntux" from charm-hub charm "ubuntu", revision 21 in channel stable on ubuntu@20.04/stable
```

## Documentation changes

@tmihoc we're now deprecating `default-series` for `juju bootstrap`, `juju add-model` and in `juju model-config`.

## Jira reference

https://warthogs.atlassian.net/browse/JUJU-2337

### CODE REVIEW

File: cmd/juju/model/config.go [hunk 1]
```diff
@@ -342,11 +342,6 @@ func (c *configCommand) setConfig(client configCommandAPI, attrs config.Attrs) e
 
 // getConfig writes the value of a single model config key to the cmd.Context.
 func (c *configCommand) getConfig(client configCommandAPI, ctx *cmd.Context) error {
-	// if certBytes != nil {
-	// 	_, _ = ctx.Stdout.Write(certBytes)
-	// 	return nil
-	// }
-
```
Comment (SimonRichardson) [inline cmd/juju/model/config.go:349]: driveby: This seems dangerous - removed.

File: core/series/base.go [hunk 2]
```diff
@@ -48,11 +48,26 @@ func ParseBaseFromString(b string) (Base, error) {
 	return Base{OS: parts[0], Channel: channel}, nil
 }
 
+// MustParseBaseFromString is like ParseBaseFromString but panics if the string
+// is invalid.
+func MustParseBaseFromString(b string) Base {
```
Comment (SimonRichardson) [inline core/series/base.go:53]: Really useful for tests.

File: tests/main.sh [hunk 3]
```diff
@@ -236,7 +236,6 @@ if [[ $# -eq 0 ]]; then
 		echo "$(red '---------------------------------------')"
 		echo ""
 		show_help
-		exit 1
```
Comment (SimonRichardson) [inline tests/main.sh:239]: driveby: fixed shellcheck

**Review by hmlanigan — COMMENTED** [review]
'default-series' is no longer showing up in the model config and bootstrap help where config The following keys are available: are shown . Which makes it difficult to describe as deprecated. Add a line in each of the help messages? 

Still working

**Review by hmlanigan — COMMENTED** [review]
QA bug found:
```
$ juju add-model two --config default-series=jammy
$ juju model-config | grep default-
default-base                       default  ""
default-series                     model    jammy
default-space                      default  ""
storage-default-filesystem-source  model    lxd
$ juju deploy ubuntu
Located charm "ubuntu" in charm-hub, revision 21
Deploying "ubuntu" from charm-hub charm "ubuntu", revision 21 in channel stable on ubuntu@20.04/stable
```

File: environs/config/config.go [hunk 4]
```diff
@@ -370,35 +374,28 @@ func (method HarvestMode) HarvestUnknown() bool {
 	return method&HarvestUnknown != 0
 }
 
-// HasDefaultSeries defines a interface if a type has a default series or not.
-type HasDefaultSeries interface {
-	DefaultSeries() (string, bool)
-}
-
 // GetDefaultSupportedLTS returns the DefaultSupportedLTS.
```
Comment (hmlanigan) [inline environs/config/config.go:377]: GetDefaultSupportedLTSBase comment block should be updated for bases rather than series

**Review by hmlanigan — COMMENTED** [review]
model migration and generate-image testing went well.

Finished with review and qa I think.

**Comment by jameinel** [comment]
I think "This means we only ever store a default-series" you mean "This means we only ever store a default-base" (just from your PR description.

File: environs/config/config.go [hunk 5]
```diff
@@ -2020,8 +2029,8 @@ var configSchema = environschema.Fields{
 		Type:        environschema.Tstring,
 		Group:       environschema.EnvironGroup,
 	},
-	DefaultSeriesKey: {
-		Description: "The default series of Ubuntu to use for deploying charms, will act like --series when deploying charms",
+	DefaultBaseKey: {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply environs/config/config.go:2032]: It's been removed completely, so not required.

File: cmd/juju/commands/bootstrap.go [hunk 6]
```diff
@@ -305,7 +306,8 @@ func (c *bootstrapCommand) SetFlags(f *gnuflag.FlagSet) {
 	c.ModelCommandBase.SetFlags(f)
 	f.StringVar(&c.ConstraintsStr, "constraints", "", "Set model constraints")
 	f.StringVar(&c.BootstrapConstraintsStr, "bootstrap-constraints", "", "Specify bootstrap machine constraints")
-	f.StringVar(&c.BootstrapSeries, "bootstrap-series", "", "Specify the series of the bootstrap machine")
+	f.StringVar(&c.BootstrapSeries, "bootstrap-series", "", "Specify the series of the bootstrap machine (deprecated use bootstrap-series)")
+	f.StringVar(&c.BootstrapBase, "bootstrap-base", "", "Specify the base of the bootstrap machine")
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply cmd/juju/commands/bootstrap.go:310]: Yeah, I'll add it to the help.

**Comment by SimonRichardson** [comment]
> I think "This means we only ever store a default-series" you mean "This means we only ever store a default-base" (just from your PR description.

Indeed, fixed.

File: cmd/juju/commands/bootstrap.go [hunk 7]
```diff
@@ -1129,8 +1159,8 @@ func (c *bootstrapCommand) controllerDataRefresher(
 
 func (c *bootstrapCommand) handleCommandLineErrorsAndInfoRequests(ctx *cmd.Context) (bool, error) {
 	if c.BootstrapImage != "" {
-		if c.BootstrapSeries == "" {
-			return true, errors.Errorf("--bootstrap-image must be used with --bootstrap-series")
+		if c.BootstrapBase == "" {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply cmd/juju/commands/bootstrap.go:1166]: This looks like an issue introduced when arch was required. I've filed a bug for now https://bugs.launchpad.net/juju/+bug/1999753

File: cmd/juju/model/config.go [hunk 8]
```diff
@@ -64,15 +64,18 @@ from one model to another:
...
 	modelConfigHelpDocKeys = `
 The following keys are available:
 `
 	modelConfigHelpDocPartTwo = `
 Examples:
 
-Print the value of default-series:
-    juju model-config default-series
+Print the value of default-base:
+    juju model-config default-base
```
Comment (hmlanigan) [inline cmd/juju/model/config.go:78]: one more `default-series` at line 93.

**Review by hmlanigan — APPROVED** [review]
LGTM, the one QA fail is resolved, a tiny nit.
